### PR TITLE
[TS] LPS-148397

### PR DESCRIPTION
**Pull request by holatuwol**
https://issues.liferay.com/browse/LPS-148397
https://issues.liferay.com/browse/PTR-2903

Based on the PTR, we've created a new SevenZipWriterImpl in order to allow creating 7z archives, which resolves the localization issue that we had with regular zip archives.

I'm sending this to core-infra first, due to the updates to the portal-kernel zip API.

The original pull request https://github.com/liferay-ts-reviewer/liferay-portal/pull/27 created a new interface SevenZipWriterFactory, but I thought it made sense to update ZipWriterFactory instead. If you feel that the original pull request approach makes more sense, please let me know.

### CODE REVIEW

**Comment by tinatian** [comment]
@holatuwol @hongvo2308  I believe the root cause is we are using non-ascii character as file/folder name, right? If this is the case, storing normalized file/folder name will solve this issue, could you please try it out? For normalizer, please see com.liferay.normalizer.Normalizer.

Tina.

**Comment by hongvo2308** [comment]
Hi @tinatian 
I have tried using the function `com.liferay.normalizer.Normalizer.normalizeToAscii` with the folder name before updating/adding the folder. But it does not solve the issue for this case.

**Comment by hongvo2308** [comment]
@tinatian 
I just added a new action on the menu for "Download as 7z", The original action "Download" still remains the same. So users can choose what activities they want to take.

**Comment by holatuwol** [comment]
Hi @tinatian , just wanted to check whether there were any additional concerns with the fix as it has been implemented.

File: portal-impl/src/com/liferay/portal/zip/ZipWriterFactoryImpl.java [hunk 1]
```diff
@@ -43,4 +45,13 @@ public ZipWriter getZipWriter(File file) {
 		return new ZipWriterImpl(file);
 	}
 
+	@Override
```
Comment (tinatian) [inline portal-impl/src/com/liferay/portal/zip/ZipWriterFactoryImpl.java:48]: @holatuwol @hongvo2308 You also need the protection for PropsValues.ZIP_FILE_WRITER_EXPORT_BUFFER_SIZE here.
As an API method, you are not sure which zip writer will be used in export process.

File: portal-kernel/src/com/liferay/portal/kernel/zip/ZipWriter.java [hunk 2]
```diff
@@ -32,28 +35,49 @@
...
+
+		addEntry(name, sb.toString());
+	}
 
 	/**
 	 * @deprecated As of Athanasius (7.3.x), replaced by {@link #getFile()}
 	 */
 	@Deprecated
-	public byte[] finish() throws IOException;
+	public default byte[] finish() throws IOException {
```
Comment (tinatian) [inline portal-kernel/src/com/liferay/portal/kernel/zip/ZipWriter.java:60]: @holatuwol Please remove deprecated methods. We are now allowed to do this.

File: portal-impl/src/com/liferay/portal/zip/LarZipWriterImpl.java [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+import java.util.AbstractMap;
+import java.util.Iterator;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * @author Minhchau Dang
+ */
+public class LarZipWriterImpl extends ZipWriterImpl {
```
Comment (tinatian) [inline portal-impl/src/com/liferay/portal/zip/LarZipWriterImpl.java:34]: @holatuwol @hongvo2308 This should not be a type of ZipWritter, it should be a wrapper of ZipWritter which will auto flush when exceeding ZIP_FILE_WRITER_EXPORT_BUFFER_SIZE.

**Comment by tinatian** [comment]
@holatuwol @hongvo2308 Please check my message for the backend logic for this issue. 

I am not sure if this is OK, please check with the Document&Media team.
> I just added a new action on the menu for "Download as 7z", The original action "Download" still remains the same. So users can choose what activities they want to take.

Tina.

**Comment by tinatian** [comment]
@holatuwol @hongvo2308 Please see my comments, I am going to close this pull, please resend if needed.

Thanks.
Tina.

**Comment by tinatian** [comment]
ci:close

**Pull request closed without merging**